casper/src/main/scala/coop/rchain/casper/util/DoublyLinkedDagOperations.scala

```diff
@@ -0,0 +1,97 @@
+package coop.rchain.casper.util
+
+import coop.rchain.casper.Estimator.BlockHash
+
+import scala.collection.immutable.{HashMap, HashSet}
+
+trait DoublyLinkedDag[A] {
+  val parentToChildAdjacencyList: Map[A, Set[A]]
+  val childToParentAdjacencyList: Map[A, Set[A]]
+  val dependencyFree: Set[A]
+}
+case class BlockDependencyDag(parentToChildAdjacencyList: Map[BlockHash, Set[BlockHash]],
+                              childToParentAdjacencyList: Map[BlockHash, Set[BlockHash]],
+                              dependencyFree: Set[BlockHash])
+    extends DoublyLinkedDag[BlockHash]
+
+object BlockDependencyDag {
+  def empty: BlockDependencyDag =
+    new BlockDependencyDag(
+      HashMap.empty[BlockHash, Set[BlockHash]],
+      HashMap.empty[BlockHash, Set[BlockHash]],
+      HashSet.empty[BlockHash]
+    )
+}
+
+object MapHelper {
+  def updatedWith[A, B](map: Map[A, B], key: A)(default: B)(f: B => B): Map[A, B] = {
+    val newValue = map.get(key).fold(default)(f)
+    map.updated(key, newValue)
+  }
+}
+
+object DoublyLinkedDagOperations {
+  def add[A](dag: DoublyLinkedDag[A], parent: A, child: A): DoublyLinkedDag[A] = {
+    val parentToChildAdjacencyList: Map[A, Set[A]] = dag.parentToChildAdjacencyList
+    val childToParentAdjacencyList: Map[A, Set[A]] = dag.childToParentAdjacencyList
+    val dependencyFree: Set[A]                     = dag.dependencyFree
+
+    val updatedParentToChildAdjacencyList =
+      MapHelper.updatedWith(parentToChildAdjacencyList, parent)(Set(child))(_ + child)
+    val updatedChildToParentAdjacencyList =
+      MapHelper.updatedWith(childToParentAdjacencyList, child)(Set(parent))(_ + parent)
+
+    val postParentDependencyFree = if (updatedChildToParentAdjacencyList.contains(parent)) {
```

Comment:
Something about these `if` statements seems off. The child necessarily depends on the parent (otherwise why are we doing the `add` operation?), so if child is in `dependencyFree` then it should be removed. And if `parent` is not the child of any other node (i.e. `updatedChildToParentAdjacencyList(parent).isEmpty`) then it can be added to `dependencyFree`